Add subclass driver

hernanmd opened the pull request:
This PR implements a first version of "Add subclass" refactoring as a driver:

https://github.com/pharo-project/pharo/assets/4825959/663878b5-19e6-4c7b-a0c8-bf097d2bab4a

## Main features

- The Spec UI to add a new subclass (see video above).
- It checks and informs interactively if the class already exists.
- It checks invalid class names.
- Default class comment could be customized implementing `blankCommentTemplate` in the class selected as superclass.

## Implementation notes

- A `RBAddNewClassRefactoring` which differs from the **Insert New Class** refactoring in that does not reparent the subclasses to the new class.
- The driver implementation in `ReAddSubclassDriver` and `ReAddSubclassDriverTest`.

## Known issues

- `CmdCommandAborted signal` and `CmCommandAborted signal` are failing right now and it should be investigated why (debugger keep opening which could mean the exception is not properly handled).
- Next version will include more tests.

Commit:
Add RBAddNewClassRefactoring and ReAddSubclassDriver
Add tests
Migrate the Syc user

pick e5f0165f76 Add RBAddNewClassRefactoring and ReAddSubclassDriver Add tests Migrate the Syc user
pick 6e9c5a4366 Add StRefactoringAddClassPresenter Update #SycCmAddSubclassCommand
pick a763b302e9 Update privateTransform so we don't reparent classes in add subclass. Add event hooks in add class presenter.

Commit:
Add StRefactoringAddClassPresenter
Update #SycCmAddSubclassCommand

Commit:
Update privateTransform so we don't reparent classes in add subclass.
Add event hooks in add class presenter.

Commit:
Refactorings
Enhance Add subclass UI

Commit:
Enable setting the class comment to the add new class refactoring from the driver.
Fix the layout so comment presenter is bigger.
Add use comment template checkbox.
Refactoring initialize presenters.
Show superclass in title dialog.
Fix updating tag items.
Temporary fix of onCancel: block using CmCommandAborted or CmdCommandAborted, both raising a Debugger when signaled.

Commit:
Remove use of ReClassChainAllowsClassInsertionCondition.

Commit:
Remove classTraits : '{} + TraitedClass'

Commit:
Remove commits from other branch accidentally included.
Restore previous versions.

pick 9b6154104e Remove classTraits : '{} + TraitedClass'
pick 3d53473f5a Remove commits from other branch accidentally included. Restore previous versions.
pick 542211943e Clean another leftover from another branch

Commit:
Clean another leftover from another branch

balsa-sarenac commented on src/Refactoring-UI/StRefactoringAddClassPresenter.class.st:
```diff
@@ -0,0 +1,234 @@
+It provides a request dialog to add a new class in the system, including presenters to specify its name, package tag and comment.
+
+"
+Class {
```
`textInput` is not used from superclass, pay attention

hernanmd replied:
Good catch, I re-used the variable from superclass.

balsa-sarenac commented on src/Refactoring-Core/RBAddNewClassRefactoring.class.st:
```diff
@@ -0,0 +1,34 @@
+"
+This refactorings differs from the Insert New Class refactoring in that does not reparent the subclasses to the new class.
+"
+Class {
+	#name : 'RBAddNewClassRefactoring',
+	#superclass : 'RBInsertNewClassRefactoring',
+	#category : 'Refactoring-Core-Refactorings',
+	#package : 'Refactoring-Core',
+	#tag : 'Refactorings'
+}
```
We can reuse `RBInsertNewClassRefactoring` without setting the subclasses like this:

```
	(RBInsertNewClassRefactoring className: newClassName)
		superclass: targetClass asString;
		packageName: targetClass package name;
		tagName: targetClass packageTag name;
		execute
```

hernanmd replied:
The purpose of AddNewClassRefactoring is that does not reparent the subclass(es) to the new class.

I understand here you want to avoid having a new class? 

In this sense I'm not sure it is better "to be clever" (it is always a mess to check which message should be avoided to have one feature active or not. I would prefer to have two explicit refactoring classes: One for adding a class with reparent (Insert), and another one without reparent (just Add).

hernanmd replied:
I will make it as resolved for now, but feel free to open a new issue or we could check it later if it makes sense to merge both classes.

balsa-sarenac replied:
Let's remove this class, since we have InsertNewClass, let's stay on track and follow the API we have so far

balsa-sarenac commented on src/SystemCommands-ClassCommands/SycCmAddSubclassCommand.class.st:
```diff
@@ -34,7 +60,78 @@ SycCmAddSubclassCommand >> name [
+	
+	refactoringScopes := context refactoringScopes.
+	targetClass := context lastSelectedClass.
+	(StRefactoringAddClassPresenter on: self)
+		onCancel: [ : dialog |
+			"CmdCommandAborted signal"
+			cancelled := true.
+			dialog close ];
+		onAccept: [ :dialog | dialog close ];
+		openModal
```
Ideally, this should be moved to the driver as well

hernanmd replied:
I'm not sure how to do it. Maybe for a next iteration? (I'm trying to not make this PR bigger)

Ducasse replied:
We should not have any logic in the command. This should be moved to the driver.

Ducasse replied:
Just move it there. 
The length of the PR is only important once we are feature complete :)

balsa-sarenac commented on src/Refactoring-UI/ReAddSubclassDriver.class.st:
```diff
@@ -0,0 +1,113 @@
+	refactoring := (RBAddNewClassRefactoring className: self subclass)
+		model: model;
+		superclass: self superclass;
+		subclasses: self superclass subclasses;
```
We can omit this line

balsa-sarenac reviewed (changes requested):
Great work! I really like the UI for adding new subclass! Left some comments inline, based on our discussion

Commit:
Fix SyntaxError

Commit:
Fix accidental commit which overwrote #applicabilityPreconditions in RBInsertNewClassRefactoring
Remove unnecessary line in ReAddSubclassDriver>>configureRefactoring
Remove unnecessary instance variable newClassPresenter in StRefactoringAddClassPresenter. Re-use 'textInput' from superclass.
Update initialExtent setting.
Remove unnecessary #example method.

Commit:
Moving out logic from Command (part 1)

hernanmd committed:
Merge branch 'pharo-project:Pharo13' into add_subclass_driver

balsa-sarenac commented on src/SystemCommands-ClassCommands/SycCmAddSubclassCommand.class.st:
```diff
@@ -5,50 +5,33 @@ Class {
+	refactoring subclass
+		ifNil: [ ^ self ]. " cancelled "
+	refactoring runRefactoring.
+	newClassName := refactoring subclass.
```
A suggestion: temp `refactoring` might be called `driver`

Commit:
Adapt tests to new changes in the driver.
Provide two instance creation helpers to provide a dialog with the superclass and its packages.
Refactor ReAddSubclassDriver methods.

Commit:
Merge e67ace6aad36205049c5040e5c01d9bae297077d

Commit:
Minor temp variable rename.

Ducasse commented:
@hernanmd can you take the point of balsa into account so that we can integrate this issue.

hernanmd committed:
Merge branch 'pharo-project:Pharo13' into add_subclass_driver

Commit:
Fix two critics:
- the following classes have unused instance variables and should be cleaned: SycCmAddSubclassCommand
- the following classes have uncategorized methods: an OrderedCollection(RBInteractionDriver)
Re-use RBInsertNewClassRefactoring also for adding new refactoring

hernanmd commented:
> @hernanmd can you take the point of balsa into account so that we can integrate this issue.

Done
